echo_test.go

```diff
@@ -17,11 +17,127 @@
 package main
 
 import (
+	"bytes"
+	"encoding/json"
+	"errors"
+	"io"
+	"net/http"
+	"net/http/httptest"
+	"strings"
 	"testing"
+	"time"
+
+	"github.com/stretchr/testify/assert"
 )
 
 const prefixURI = "/v0"
 
 func TestInvalidMethod(t *testing.T) {
 
+	req, err := http.NewRequest("GET", "/v0/echo", nil)
+
+	if err != nil {
+		t.Error(err)
+		return
+	}
+
+	rr := httptest.NewRecorder()
+	handler := EchoHandler{}
+	handler.ServeHTTP(rr, req)
+
```

Comment:
Part of the `EchoHandler` contract is that for errors, it returns an [rfc7807](https://datatracker.ietf.org/doc/html/rfc7807)-formatted JSON object (which we're encoding with https://github.com/mschneider82/problem).  This is so that it's possible to pass along more detailed failure information to clients (in a structured manner). So, for an invocation like this that we expect to return an HTTP status of 405 (`http.StatusMethodNotAllowed`), the body of the response should also contain JSON looking like:

```json
{
  "status": 405,
  "detail": "Method not allowed"
}
```

So, to test that part of the content, we could unmarshal the body to a struct in order to validate that it is in fact JSON (formatted according to rfc7807), and that it at least minimally contains a `status` attribute that is also 405.

You'd need to create a struct for this purpose, and it only need contain the attributes we plan to test for (anything else there will just be silently ignored).

```golang
type httperr struct {
    Status int32
}
```

Then create an instance of the struct and unmarshal it, just like you did when testing that a successful response returned a valid `Echo` object.

```golang
e := httperr{}
err := json.Unmarshal(body, &e)
assert.Equal(t, http.StatusMethodNotAllowed, e.status, ...)
```

The same applies to the other functions that test for errors.